lib/ddtrace/workers/trace_writer.rb

```diff
@@ -0,0 +1,199 @@
+require 'ddtrace/logger'
+require 'ddtrace/transport/http'
+
+require 'ddtrace/event'
+require 'ddtrace/worker'
+require 'ddtrace/workers/polling'
+require 'ddtrace/workers/queue'
+
+module Datadog
+  module Workers
+    # Writes traces to transport synchronously
+    class TraceWriter < Worker
+      attr_reader \
+        :transport
+
+      def initialize(options = {})
+        transport_options = options.fetch(:transport_options, {})
+
+        if transport_options.is_a?(Proc)
+          transport_options = { on_build: transport_options }
+        end
+
+        transport_options[:hostname] = options[:hostname] if options.key?(:hostname)
+        transport_options[:port] = options[:port] if options.key?(:port)
+
+        @transport = options.fetch(:transport) do
+          Transport::HTTP.default(transport_options)
+        end
+      end
+
+      def perform(traces)
+        write_traces(traces)
+      end
+
+      def write(trace)
+        write_traces([trace])
+      end
+
+      def write_traces(traces)
+        traces = process_traces(traces)
+        flush_traces(traces)
+      rescue StandardError => e
+        Datadog::Logger.log.error(
+          "Error while writing traces: dropped #{traces.length} items. Cause: #{e} Location: #{e.backtrace.first}"
+        )
+      end
+
+      def process_traces(traces)
+        # Run traces through the processing pipeline
+        traces = Pipeline.process!(traces)
+
+        # Inject hostname if configured to do so
+        inject_hostname!(traces) if Datadog.configuration.report_hostname
+
+        traces
+      end
+
+      def flush_traces(traces)
+        transport.send_traces(traces).tap do |response|
+          flush_completed.publish(response)
+        end
+      end
+
+      def inject_hostname!(traces)
+        traces.each do |trace|
+          next if trace.first.nil?
+
+          hostname = Datadog::Runtime::Socket.hostname
+          unless hostname.nil? || hostname.empty?
+            trace.first.set_tag(Ext::NET::TAG_HOSTNAME, hostname)
+          end
+        end
+      end
+
+      def flush_completed
+        @flush_completed ||= FlushCompleted.new
+      end
+
+      # Flush completed event for worker
+      class FlushCompleted < Event
+        def initialize
+          super(:flush_completed)
+        end
+
+        def publish(response)
+          super(response)
+        end
+      end
+    end
+
+    # Writes traces to transport asynchronously,
+    # using a thread & buffer.
+    class AsyncTraceWriter < TraceWriter
```

Comment:
Should I move this to its own file? Was useful seeing both of these together, but it might make sense to have one class per file whenever possible/reasonable.

Reply:
I don't see a problem with having them both in this file right now, specially since one inherits from the other, making navigating the implementation of `AsyncTraceWriter` easier.

If sync and async implementation become more divergent we could move them to separate files with little hassle.